fusillade/resource.py

```diff
@@ -0,0 +1,225 @@
+"""
+## ResourceType
+The ResourceType class is an abstraction of a node in cloud directory that represents a type of resource we would like
+to apply access control logic (ACL).
+
+### Actions
+This type of resource has certain action that can be performed on or with it, and these actions are store in an
+attribute called `actions`. Any access policy associated with this resource type must only include actions that the
+resource supports. This is checked before new policies are added, and nonconforming access policies are rejected. If an
+actions is removed from the resource type all existing access policies with this action will be removed.
+
+### Owner policy
+The owner policy is added to the resource type and is used to determine what actions the owner of a resource Id can
+perform on a resource.
+"""
+
+import json
+import os
+from typing import List, Dict, Any, Union
+
+from fusillade.clouddirectory import CloudNode, cd_client, ConsistencyLevel, logger
+from fusillade.errors import FusilladeHTTPException, FusilladeBadRequestException
+from fusillade.policy.validator import verify_iam_policy
+
+proj_path = os.path.dirname(__file__)
+default_resource_owner_policy = os.path.join(proj_path, '..', 'policies', 'default_resource_owner_policy.json')
+
+
+class ResourceType(CloudNode):
+    _attributes = ['actions'] + CloudNode._attributes
+    _facet: str = 'NodeFacet'
+    object_type: str = 'resource'
+    _default_policy_path: str = default_resource_owner_policy
+    policy_type = 'Resource'
+
+    def __init__(self, *args, **kwargs):
+        super(ResourceType, self).__init__(*args, **kwargs)
+        self._actions: str = None
+
+    @classmethod
+    def create(cls,
+               name: str,
+               actions: List[str],
+               owner_policy: Dict[str, Any] = None,
+               creator: str = None,
+               **kwargs) -> 'ResourceType':
+        """
```

Comment:
This is a long and complex method. For comprehensibility, it may be useful to separate logic managing batched writes from logic managing the creation of CD nodes (which I don't completely understand). Context managers can be useful for these cases.

Reply:
Will look into implementing in a future PR